src/protocols/secure_channel/PASESession.cpp

```diff
@@ -578,115 +651,131 @@ CHIP_ERROR PASESession::HandleMsg1_and_SendMsg2(const System::PacketBufferHandle
+        SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+        SuccessOrExit(err = tlvWriter.Finalize(&msg3));
+
+        mNextExpectedMsg = MsgType::StatusReport;
 
         // Call delegate to send the Msg3 to peer
-        err = mExchangeCtxt->SendMessage(Protocols::SecureChannel::MsgType::PASE_Spake2p3, bbuf.Finalize());
+        err = mExchangeCtxt->SendMessage(MsgType::PASE_Spake2p3, std::move(msg3), SendFlags(SendMessageFlags::kExpectResponse));
         SuccessOrExit(err);
     }
 
     ChipLogDetail(SecureChannel, "Sent spake2p msg3");
 
-    mPairingComplete = true;
-
-    // Forget our exchange, as no additional messages are expected from the peer
-    mExchangeCtxt = nullptr;
-
-    // Call delegate to indicate pairing completion
-    mDelegate->OnSessionEstablished();
-
 exit:
 
     if (err != CHIP_NO_ERROR)
     {
-        SendErrorMsg(spake2pErr);
+        SendStatusReport(spake2pErr);
     }
     return err;
 }
 
-CHIP_ERROR PASESession::HandleMsg3(const System::PacketBufferHandle & msg)
+CHIP_ERROR PASESession::HandleMsg3(System::PacketBufferHandle && msg)
 {
-    CHIP_ERROR err              = CHIP_NO_ERROR;
-    const uint8_t * hash        = msg->Start();
-    Spake2pErrorType spake2pErr = Spake2pErrorType::kUnexpected;
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    uint16_t spake2pErr = kProtocolCodeInvalidParam;
 
     ChipLogDetail(SecureChannel, "Received spake2p msg3");
 
-    // We will set NextExpectedMsg to PASE_Spake2pError in all cases
+    // We will set NextExpectedMsg to StatusReport in all cases
     // However, when we are using IP rendezvous, we might set it to PASE_Spake2p1.
-    mNextExpectedMsg = Protocols::SecureChannel::MsgType::PASE_Spake2pError;
+    mNextExpectedMsg = MsgType::StatusReport;
+
+    System::PacketBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    const uint8_t * peer_verifier;
+    size_t peer_verifier_len = 0;
 
-    VerifyOrExit(hash != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
-    VerifyOrExit(msg->DataLength() == kMAX_Hash_Length, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
 
-    err = mSpake2p.KeyConfirm(hash, kMAX_Hash_Length);
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == 1, err = CHIP_ERROR_INVALID_TLV_TAG);
+    peer_verifier_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(peer_verifier));
+
+    VerifyOrExit(peer_verifier_len == kMAX_Hash_Length, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    err = mSpake2p.KeyConfirm(peer_verifier, peer_verifier_len);
     if (err != CHIP_NO_ERROR)
     {
-        spake2pErr = Spake2pErrorType::kInvalidKeyConfirmation;
+        spake2pErr = kProtocolCodeNoSharedRoot;
```

Comment:
This is spec-compliant but makes no sense. I think we made a quick mistake in the spec:

> . Verify `Pake3.cA` against `cA`. If verification fails, the responder SHALL send a <<ref_SC_StatusReports, status report>>: `StatusReport(GeneralCode: FAILURE, ProtocolId: <<bookmark-protocol-id-secure-channel, SECURE_CHANNEL>>, ProtocolCode: <<ref_NO_SHARED_TRUST_ROOTS>>)` and perform no further processing.

@balducci-apple is there a different status we could use here?

Reply:
We should be using `ref_INVALID_PARAMETER`

Reply:
So that means we use INVALID_PARAM in every error condition in PASE. That's ok. Just want to confirm.
Earlier we had a different error value for KeyConfirmation failure.

Reply:
Updated the code to use INVALID_PARAM. If the spec changes to use a different error code, we can update the SDK to match it in a follow on PR.